feat: add CJK friendly emphasis extension

3w36zj6 opened the pull request:
CommonMark has a problem that the following emphasis marks `**` are not recognized as emphasis marks in CJK.

```md
**このアスタリスクは強調記号として認識されず、そのまま表示されます。**この文のせいで。

**该星号不会被识别，而是直接显示。**这是因为它没有被识别为强调符号。

**이 별표는 강조 표시로 인식되지 않고 그대로 표시됩니다(이 괄호 때문에)**이 문장 때문에.
```

This pull request introduces support for CJK-friendly emphasis handling in the Markdown parser, aligning with the [CommonMark CJK-friendly amendments specification](https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md). 

It adds a new option to enable CJK-friendly emphasis parsing, updates the delimiter run logic to properly handle CJK characters and punctuation, and includes comprehensive tests to verify the new behavior. By default, the feature is disabled to maintain backward compatibility.

In addition to the specification, I also refer to the [Tips for Implementers](https://github.com/tats-u/markdown-cjk-friendly/blob/main/implementers-tips.md) and [Concrete ranges of each terms](https://github.com/tats-u/markdown-cjk-friendly/blob/main/ranges.md) in [tats-u/markdown-cjk-friendly](https://github.com/tats-u/markdown-cjk-friendly) for implementation.

3w36zj6 committed:
feat: add CJK friendly emphasis extension

ollpu commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2248,6 +2252,36 @@ fn fixup_end_of_definition_list(tree: &mut Tree<Item>, list_ix: TreeIndex) {
+        prev = Some(ch);
+    }
+    (prev, prev_prev)
+}
```
Related, maybe this should not take `ix`, so that it's the caller's responsibility to slice the string.

tats-u replied:
.char() returns DoubleEndedIterator.

https://doc.rust-lang.org/src/core/str/iter.rs.html#125
https://stdrs.dev/nightly/x86_64-pc-windows-gnu/src/core/str/validations.rs.html#79-113

3w36zj6 committed:
fix: remove unnecessary inline annotations

3w36zj6 committed:
perf: optimize utility for retrieving last two characters by reversing the iterator

3w36zj6 committed:
refactor: accept prefix string directly for character retrieval

tats-u commented on pulldown-cmark/tests/html.rs:
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
+#[test]
+fn cjk_friendly_emphasis_simplified_chinese_punctuation() {
+    let input = "这是。**强调**吗？";
+    let expected = "<p>这是。<strong>强调</strong>吗？</p>\n";
```
This is not specific to this extension.

tats-u commented on pulldown-cmark/tests/html.rs:
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
+#[test]
+fn cjk_friendly_emphasis_handles_cjk_suffix() {
+    let input = "_強調_箇所";
+    let expected = "<p><em>強調</em>箇所</p>\n";
```
This cannot be fixed even by this extension.

tats-u commented on pulldown-cmark/tests/html.rs:
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
+#[test]
+fn cjk_friendly_emphasis_enabled() {
+    let input = "これは_強調_です";
+    let expected = "<p>これは<em>強調</em>です</p>\n";
```
This cannot be fixed even by this extension.

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2398,23 +2424,44 @@ fn delim_run_can_open(
         return true;
     }
-    let prev_char = s[..ix].chars().last().unwrap();
+    let (prev_char_opt, prev_prev_char) = previous_two_chars(&s[..ix]);
```
Don't have pulldown-cmark retrieve `prev_prev_char` unless it's actually needed in this extension (especially when it has not been turned on)

tats-u commented on pulldown-cmark/src/cjk.rs:
```diff
@@ -0,0 +1,87 @@
+
+/// Returns true when `ch` is considered a CJK character by the
+/// CommonMark CJK-friendly amendments specification.
+pub(crate) fn is_cjk_character(ch: char) -> bool {
```
Add Unicode version. If it is lower than 17, update the code to 17.

3w36zj6 committed:
Merge remote-tracking branch 'origin/master' into feature/add-cjk-friendly-emphasis-extension

tats-u commented on pulldown-cmark/tests/html.rs:
```diff
@@ -316,6 +316,88 @@ fn trim_space_nl_at_end_of_paragraph() {
+
+    assert_eq!(expected, output);
+}
+
```
The current test cases are currently redundant and far from sufficient. You need to add edge cases (e.g. ones using variation selectors or decomposed Hangul parts) and cases where an `*`/`**` is placed between an alphabet and a CJK punctuation character.

tats-u commented on pulldown-cmark/src/lib.rs:
```diff
@@ -768,6 +769,12 @@ bitflags::bitflags! {
+        ///
+        /// Aligns delimiter run handling with the
+        /// [CommonMark CJK-friendly amendments](https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md),
+        /// enabling emphasis between CJK characters without surrounding whitespace.
```
```suggestion
        /// [CommonMark CJK-friendly amendments/extension](https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md),
        /// enabling emphasis around CJK characters without surrounding whitespace.
```

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2472,18 +2523,32 @@ fn delim_run_can_close(
+            || is_cjk_character(next_char)
+    } else {
+        true
+    }
```
```suggestion
    next_char.is_whitespace() || if !cjk_friendly {
        is_punctuation(next_char)
    } else {
        !prev_non_cjk_seq
            || is_non_cjk_punctuation_character(next_char)
            || is_cjk_character(next_char)
    }
```

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2428,16 +2460,31 @@ fn delim_run_can_open(
     prev_char.is_whitespace()
-        || is_punctuation(prev_char) && (delim != b'\'' || ![']', ')'].contains(&prev_char))
+        || (prev_non_cjk_seq && (delim != b'\'' || !blocked_by_quote))
+        || (cjk_friendly && (prev_cjk_seq || prev_is_ivs))
```
Utilize `if` as expression, block expressions, and `&&`/`||` operators to reduce overhead when this extension is turned off

tats-u reviewed (changes requested):
The current change is still far from official recognition. While I haven't thoroughly understood the pulldown-cmark codebase yet, there's still room for improvement—including areas beyond what I've pointed out—and the test cases are at best inadequate.

tats-u commented:
pulldown-cmark is one of the fastest Markdown parsers, so we need to keep the performance when this extension is turned off. We may have to introduce CodSpeed to monitor performance regression in advance.

3w36zj6 committed:
docs: clarify description of CJK-friendly emphasis flag

Co-Authored-By: tats-u <12870451+tats-u@users.noreply.github.com>

3w36zj6 committed:
refactor: simplify logic for CJK-friendly delimiter handling

Co-Authored-By: tats-u <12870451+tats-u@users.noreply.github.com>

3w36zj6 committed:
docs: specify Unicode version for CJK character ranges

3w36zj6 committed:
test: add cjk friendly emphasis specification

3w36zj6 committed:
test: regenerate suites

3w36zj6 committed:
fix: handle variation sequences in cjk friendly emphasis

3w36zj6 committed:
perf: defer cjk friendly emphasis lookbehind

3w36zj6 committed:
perf: skip cjk emphasis classification when disabled

3w36zj6 commented:
Thanks for the review.

For the tests, I changed them to follow the same pattern as the other extensions: the suite is generated from a spec fixture. The source URLs I used are listed explicitly with commit hashes. While doing that, I also fixed the boundary checks for CJK-friendly emphasis cases involving variation selectors. The test coverage should now be much more comprehensive.

For the performance concern, I updated the implementation so that, when CJK-friendly emphasis is disabled, the extra work for this extension is avoided as much as possible.

3w36zj6 committed:
test: cover cjk friendly emphasis integration behavior

3w36zj6 committed:
test: cover cjk underscore intraword restrictions

3w36zj6 committed:
fix: apply cjk-friendly flanking and underscore rules

3w36zj6 committed:
feat: add CJK-friendly emphasis CLI option

3w36zj6 committed:
docs: clarify CJK-friendly emphasis spec licenses

3w36zj6 committed:
perf: avoid repeated CJK-friendly emphasis option checks

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2523,31 +2487,105 @@ fn delim_run_can_close(
+        let _ = previous_chars.next();
+        return cjk_friendly_delim_run_can_close(
+            prev_char,
+            previous_chars.next(),
```
Nitpick/perf: this character is actually not used in most cases. I wonder if `LazyCell` is too slow (has too much overhead).

3w36zj6 committed:
perf: defer CJK emphasis lookbehind

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2488,11 +2488,13 @@ fn delim_run_can_close(
+                let mut previous_chars = s[..ix].chars().rev();
+                let _ = previous_chars.next();
+                previous_chars.next()
+            },
```
```suggestion
            || s[..(ix - prev_char.len_utf8())].chars().rev(),
```

tats-u commented on pulldown-cmark/src/cjk.rs:
```diff
@@ -0,0 +1,151 @@
+/// The source is https://github.com/tats-u/markdown-cjk-friendly/blob/ee93f3c2dffc8c7eb25c7a1c9f3f962279ac247d/ranges.md.
+pub(crate) fn is_cjk_character(ch: char) -> bool {
+    let cp = ch as u32;
+    matches!(
```
I prefer a fast path:

```suggestion
    if cp < 0x1100 {
        return false;
    }
    matches!(
```

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2472,6 +2487,18 @@ fn delim_run_can_close(
         }
     }
     let delim = suffix.bytes().next().unwrap();
+    if cjk_friendly_emphasis && matches!(delim, b'*' | b'_') {
```
`~~` should be CJK friendly too.

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2417,6 +2428,9 @@ fn delim_run_can_open(
             return false;
         }
     }
+    if cjk_friendly_emphasis && matches!(delim, b'*' | b'_') {
```
`~~` should be CJK friendly too.

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2486,6 +2513,84 @@ fn delim_run_can_close(
+        can_open,
+        can_close,
+    }
+}
```
I updated the "implementers tips":

- https://github.com/tats-u/markdown-cjk-friendly/blob/main/implementers-tips.md#flowchart-to-suppress-extra-processes
- https://github.com/tats-u/markdown-cjk-friendly/blob/main/implementers-tips.md#logical-expressions-transformations

You don't need to use the results whether characters are "non-CJK punctuation". Instead of that  you can check whether they are CJK characters or punctuation.

3w36zj6 committed:
refactor: simplify previous character lookup

Co-Authored-By: tats-u <12870451+tats-u@users.noreply.github.com>

3w36zj6 committed:
perf: add fast path for non-CJK code points

Co-Authored-By: tats-u <12870451+tats-u@users.noreply.github.com>

3w36zj6 committed:
test: apply CJK-friendly flanking to strikethrough

3w36zj6 committed:
fix: apply CJK-friendly flanking to strikethrough

3w36zj6 committed:
perf: avoid extra flanking classification

3w36zj6 committed:
refactor: use `next_back` for previous character lookup

tats-u commented on pulldown-cmark/specs/cjk_friendly_emphasis.txt:
```diff
@@ -0,0 +1,474 @@
+
+Here I strike out an exclamation point~~!~~.
+.
+<p>真<del>（她</del></p>
+<p><del>真，</del>她</p>
+<p>This<del>is</del>stricken</p>
+<p>Here I strike out an exclamation point<del>!</del>.</p>
```
https://tats-u.github.io/markdown-cjk-friendly/?sc8=55yffn7vvIjlpbl-fgoKfn7nnJ_vvIx-fuWluQoKVGhpc35-aXN-fnN0cmlja2VuCgpIZXJlIEkgc3RyaWtlIG91dCBhbiBleGNsYW1hdGlvbiBwb2ludH5-IX5-Lg&gfm=1&engine=micromark&ver=2.0.1

```suggestion
<p>真<del>（她</del></p>
<p><del>真，</del>她</p>
<p>This<del>is</del>stricken</p>
<p>Here I strike out an exclamation point~~!~~.</p>
```

tats-u commented on pulldown-cmark/tests/suite/cjk_friendly_emphasis.rs:
```diff
@@ -432,6 +432,25 @@ _abc_漢
+    let expected = r##"<p>真<del>（她</del></p>
+<p><del>真，</del>她</p>
+<p>This<del>is</del>stricken</p>
+<p>Here I strike out an exclamation point<del>!</del>.</p>
```
https://tats-u.github.io/markdown-cjk-friendly/?sc8=55yffn7vvIjlpbl-fgoKfn7nnJ_vvIx-fuWluQoKVGhpc35-aXN-fnN0cmlja2VuCgpIZXJlIEkgc3RyaWtlIG91dCBhbiBleGNsYW1hdGlvbiBwb2ludH5-IX5-Lg&gfm=1&engine=micromark&ver=2.0.1

```suggestion
<p>Here I strike out an exclamation point~~!~~.</p>
```

tats-u commented:
`~~` behaves like `**`, but both maintain traditional non-CJK-friendly behavior when wrapping non-CJK punctuation within non-CJK letters.

tats-u commented:
> This branch has conflicts that must be resolved

3w36zj6 committed:
Merge remote-tracking branch 'origin/master' into feature/add-cjk-friendly-emphasis-extension

# Conflicts:
#	pulldown-cmark/src/lib.rs
#	pulldown-cmark/src/main.rs

3w36zj6 committed:
test: align cjk-friendly strikethrough expectation

Co-Authored-By: tats-u <12870451+tats-u@users.noreply.github.com>

3w36zj6 committed:
fix: apply CJK-friendly flanking to strikethrough openers

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2620,7 +2616,7 @@ fn cjk_friendly_delim_run_flanking(
             || matches!(prev_char, '*' | '_'));
 
-    let can_open = if delim == b'*' {
+    let can_open = if matches!(delim, b'*' | b'~') {
```
```suggestion
    let can_open = if cjk_changes_flanking {
```

tats-u commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -2598,7 +2594,7 @@ fn cjk_friendly_delim_run_flanking(
 
     let prev_sequence = classify_preceding_cjk_friendly_sequence(prev_char, get_prev_prev_char);
-    let cjk_changes_flanking = delim == b'*';
+    let cjk_changes_flanking = matches!(delim, b'*' | b'~');
```
@ollpu Do you think there are delimiter marks other than emphasis (`_` / `*`) and strikethrough (`~`) that behave like `_`?

If not, this should be changed to:

```suggestion
    let cjk_changes_flanking = delim != b'_';
```

notriddle replied:
Is it even possible for this to evaluate to false here?

The `is_cjk_friendly_delim` function only returns true if `delim` is `*`, `_`, or `~`. On line 2544, we early return from this function when `delim` is `_` or `~`. That should mean that `delim` will always be `*` at this point.

tats-u replied:
You're right. `cjk_changes_flanking` is always `true`.

3w36zj6 committed:
refactor: reuse CJK flanking flag for delimiter opening

Co-Authored-By: tats-u <12870451+tats-u@users.noreply.github.com>

tats-u reviewed (approved):
The latter part seems adequate as an initial implementation of this extension.

tats-u commented:
@Martin1887 Do you have time to check this PR? I, the original author of this extension, believe the content is fine.

Martin1887 commented:
Thanks for the reminder. It is a bit quite large, but I will try to review it in the following days. I also add @notriddle as reviewer, who will find more technical details and improvements in case they exist.

tats-u commented:
Thank you for your reply. I hope playgrounds will help your review:

Official playground: https://tats-u.github.io/markdown-cjk-friendly/ (marked plugin seems to have poorer quality than those for the others)
Comrak (GLFM) playground: https://gitlab-org.gitlab.io/ruby/gems/gitlab-glfm-markdown/

notriddle commented on pulldown-cmark/src/firstpass.rs:
```diff
@@ -32,6 +36,7 @@ pub(crate) fn run_first_pass(text: &str, options: Options) -> (Tree<Item>, Alloc
         last_line_blank: false,
         allocs: Allocations::new(),
         options,
+        cjk_friendly_emphasis: options.contains(Options::ENABLE_CJK_FRIENDLY_EMPHASIS),
```
If a function already accepts `options: Options`, you should not add a `cjk_friendly_emphasis: bool` parameter.

Functions that accept `bool`ean parameters for extensions, like `scan_paragraph_interrupt_no_table`, do not have an `options: Options` parameter.

One or the other. Not both.

3w36zj6 committed:
refactor: derive CJK-friendly emphasis from parser options

notriddle committed:
Code cleanups

notriddle commented:
I found a few strikethrough-related problems, and pushed the fixes to a new commit that you can pull in if they look good.

https://github.com/notriddle/pulldown-cmark/commit/4c0777a30794d566359ced599852d0ca7c68029b

tats-u commented:
It seems this PR or your suggestion has made some changes to the single tilde syntax. While I completely agree that the single tilde syntax should be less likely to produce strikethrough effects compared to the double tilde syntax, it will also have to be reflected in my original specifications (https://github.com/tats-u/markdown-cjk-friendly/blob/main/specification.md).

tats-u commented:
From my understanding, your proposal is to change the single tilde syntax to match the conventional CJK-unfriendly asterisk syntax. Is that correct?

tats-u commented:
I know the combination of CJK-friendly extension and the single tilde syntax causes some problems by https://github.com/vercel/streamdown/issues/545.

notriddle commented:
The way it’s supposed to work in pulldown-cmark is described here: https://pulldown-cmark.github.io/pulldown-cmark/specs/strikethrough.html

> From my understanding, your proposal is to change the single tilde syntax to match the conventional CJK-unfriendly asterisk syntax. Is that correct?

Single tilde strikeout in pulldown-cmark is supposed to act likes underscores, not asterisks. In CJK-friendly mode, single tilde strikeout is supposed to act like CJK-friendly underscores.

tats-u commented:
> Single tilde strikeout in pulldown-cmark is supposed to act likes underscores, not asterisks.

Ah, I see.

tats-u commented:
I don't see any issues with merging https://github.com/notriddle/pulldown-cmark/commit/4c0777a30794d566359ced599852d0ca7c68029b. Go ahead and do it.

3w36zj6 committed:
refactor: remove redundant CJK flanking branch

guoliu commented:
An adopter data point, and a thank-you 🙏

[moss](https://github.com/Symbiosis-Lab/moss-releases) (a Rust + Tauri static-site generator & editor) hit exactly this on a published Chinese-language article: `**中文。**后面` — and every similar sentence — rendered as literal `**`, because the closing delimiter falls between CJK punctuation and an ideograph with no ASCII space between them.

We validated this PR against our real content and pipeline. With `Options::ENABLE_CJK_FRIENDLY_EMPHASIS`:

- all the affected spans render correctly, and
- our existing CommonMark/ASCII emphasis tests are unchanged — backward-compatible, exactly as the [`tats-u/markdown-cjk-friendly`](https://github.com/tats-u/markdown-cjk-friendly) spec promises.

We're shipping it now via a `[patch.crates-io]` fork of this branch, and would love to see it land and released so we can drop the patch. Happy to help test on real CJK corpora. Thanks @3w36zj6 for implementing the spec, and to the maintainers for reviewing it.

Martin1887 commented:
Please fix the errors in tests detected in CI before merging this pull request. Thanks!

3w36zj6 committed:
Merge remote-tracking branch 'origin/master' into feature/add-cjk-friendly-emphasis-extension

3w36zj6 committed:
test: isolate CJK-friendly emphasis from strikethrough
